src/OutgoingCallMessage.ts

```diff
@@ -0,0 +1,368 @@
+   */
+  addListener(
+    failureEvent: OutgoingCallMessage.Event.Failure,
+    listener: OutgoingCallMessage.Listener.Failure
+  ): this;
+  /** {@inheritDoc (OutgoingCallMessage:interface).(addListener:1)} */
+  on(
+    failureEvent: OutgoingCallMessage.Event.Failure,
+    listener: OutgoingCallMessage.Listener.Failure
+  ): this;
+
+  /**
+   * Sent event. Raised when outgoingCallMessage is sent.
+   * @example
+   * ```typescript
+   * outgoingCallMessage.addListener(OutgoingCallMessage.Event.Sent, () => {
+   *    // outgoingCallMessage sent
+   * })
+   * ```
+   *
+   * @param sentEvent - The raised event string.
+   * @param listener - A listener function that will be invoked when the event
+   * is raised.
+   * @returns - The callMessage object
+   */
+  addListener(
+    sentEvent: OutgoingCallMessage.Event.Sent,
+    listener: OutgoingCallMessage.Listener.Sent
+  ): this;
+  /** {@inheritDoc (OutgoingCallMessage:interface).(addListener:2)} */
+  on(
+    sentEvent: OutgoingCallMessage.Event.Sent,
+    listener: OutgoingCallMessage.Listener.Sent
+  ): this;
+
+  /**
+   * Generic event listener typings.
+   * @param outgoingCallMessageEvent - The raised event string.
+   * @param listener - A listener function that will be invoked when the event
+   * is raised.
+   * @returns - The outgoingCallMessage object.
+   */
+  addListener(
+    outgoingCallMessageEvent: OutgoingCallMessage.Event,
+    listener: OutgoingCallMessage.Listener.Generic
+  ): this;
+  /**
+   * {@inheritDoc (OutgoingCallMessage:interface).(addListener:3)}
+   */
+  on(
+    outgoingCallMessageEvent: OutgoingCallMessage.Event,
+    listener: OutgoingCallMessage.Listener.Generic
+  ): this;
+}
+
+/**
+ * Provides access to information about a outgoingCallMessage, including the call
+ * message content, contentType, messageType, and voiceEventSid
+ *
+ * @remarks
+ * Note that the outgoingCallMessage information is fetched as soon as possible from the
+ * native layer, but there is no guarantee that all information is immediately
+ * available. Methods such as `OutgoingCallMessage.getContent` or `OutgoingCallMessage.getSid`
+ * may return `undefined`.
+ *
+ * As outgoingCallMessage events are received from the native layer, outgoingCallMessage information will
+ * propagate from the native layer to the JS layer and become available.
+ * Therefore, it is good practice to read information from the outgoingCallMessage after an
+ * event occurs, or as events occur.
+ *
+ * - See the {@link (OutgoingCallMessage:namespace).Event} enum for events emitted by `OutgoingCallMessage`
+ *   objects.
+ * - See the {@link (OutgoingCallMessage:interface) | OutgoingCallMessage interface} for overloaded event listening
+ *   metods.
+ * - See the {@link (OutgoingCallMessage:namespace) | OutgoingCallMessage namespace} for types and enumerations
+ *   used by this class.
+ *
+ * @public
+ */
+export class OutgoingCallMessage extends EventEmitter implements CallMessage {
```

Comment:
Our original purpose for extending CallMessage is to avoid duplication of code. In this implementation, we are duplicating all of CallMessage members. I understand we also need to inherit from EventEmitter and multiple inheritance is tricky in typescript as it doesn't support it natively. There are multiple ways to address this but I'm hesitant to implement something complex for this simple use case. I'm ok if we compromise for something like: CallMessage inherits from EventEmitter. That way, OutgoingCallMessage can inherit CallMessage properly without duplication of code. Downside is, CallMessage is now an EventEmitter. I don't see any negative side effects though. Thoughts @mhuynh5757 @kpchoy ?

Reply:
I think @mhuynh5757 was concerned that a User might add a listener to `CallMessage` even though it doesn't have any listeners configured. I do see issue with duplicate code. @mhuynh5757 did you have any other points?

Reply:
I don't think Typescript mixins are overly complex, should we give it a shot and see how it looks? @kpchoy what do you think? I don't think it will take more than half an hour of experimentation but it might be throwaway work if we decide it's not worth it.

Reply:
I've seen instances where mixins become unnecessarily complex but if it takes only this much time to experiment, feel free to go ahead.

Reply:
Also, mixins is only one option. Feel free to do a bit of exploration and choose what you think fits our needs.

Reply:
Ok Ill give mixins a shot

Reply:
After review of mixins and some experimentation I think it is too complex for our current implementation. I changed it back to the original way Charlie suggested: `CallMessage inherits from EventEmitter. That way, OutgoingCallMessage can inherit CallMessage properly without duplication of code.`. Although it is still a possibility for customers to mistakenly add listeners to CallMessage, I think the possibility is slim.

Reply:
I was leaning more towards to this solution as well. I'm not too concerned about customers adding listeners to CallMessage because we are not documenting any events in the public API docs. Thoughts @mhuynh5757 ?

Reply:
Agreed. I think we should just do the `CallMessage extends EventEmitter` route. I tried some shenanigans with typing out the inherited methods on just the root `CallMessage` class and using `never` but it's not worth it.

Reply:
It won't be the end of the world if developers add useless listeners to the root `CallMessage` object.